services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/CommandRouterServiceImpl.java

```diff
@@ -207,55 +219,103 @@ private Duration getSanitizedLifespan(final Duration lifespan) {
-                    .map(factory -> factory.createCommandConsumer(tenantId, span.context()))
-                    .onSuccess(ok -> {
-                        span.log("successfully created command consumer");
-                        processedTenants.add(tenantId);
-                    })
-                    .onFailure(t -> {
-                        TracingHelper.logError(span, "failed to create command consumer", t);
-                        if (t instanceof ServerErrorException) {
-                            // add to end of queue in order to retry at a later time
-                            span.log("marking tenant for later re-try to create command consumer");
-                            tenantsToEnable.add(tenantId);
-                        }
-                    })
-                    .onComplete(r -> {
-                        span.finish();
-                    });
+                context.owner().setTimer(delay, tid -> activateCommandRouting(attempt, tracingContext));
             }
-            processTenantQueue(processedTenants, parentSpan);
-        });
+        }
+    }
+
+    private long calculateDelayMillis(final int attemptNo) {
+        if (attemptNo == 1) {
+            // run first attempt immediately
+            return 0L;
+        } else if (attemptNo > 6) {
+            // cap delay at 10secs
+            return 10000L;
+        } else {
+            // wait for 2^attemptNo times 100 millis before retrying
+            return (1 << attemptNo) * 100L;
+        }
+    }
+
+    private void activateCommandRouting(
+            final Pair<String, Integer> attempt,
+            final SpanContext tracingContext) {
+
+        if (!running.get()) {
+            // component has been stopped, no need to create command consumer in this case
+            tenantsInProcess.remove(attempt.one());
+            return;
+        }
+
+        final Span span = tracer.buildSpan("re-enable command routing for tenant")
+                .addReference(References.FOLLOWS_FROM, tracingContext)
+                .withTag(TracingHelper.TAG_TENANT_ID, attempt.one())
+                .start();
+        final var logEntries = new HashMap<String, Object>(2);
+        logEntries.put("attempt#", attempt.two());
+        tenantClient.get(attempt.one(), span.context())
+            .map(tenantObject -> commandConsumerFactories.getClient(tenantObject))
+            .map(factory -> factory.createCommandConsumer(attempt.one(), span.context()))
+            .onSuccess(ok -> {
+                logEntries.put(Fields.MESSAGE, "successfully created command consumer");
+                span.log(logEntries);
+                reenabledTenants.add(attempt.one());
+            })
+            .onFailure(t -> {
+                logEntries.put(Fields.MESSAGE, "failed to create command consumer");
+                logEntries.put(Fields.ERROR_OBJECT, t);
+                TracingHelper.logError(span, logEntries);
```

Comment:
`TracingHelper.logError(Span, Map)` sets the *error* tag to `true`.

Reply:
Yes, I think that should be done here. The `re-enable command routing for tenant` action failed so the corresponding span gets the error tag.

Reply:
my point is that there is nothing that needs to be changed because the span is already marked as erroneous with the current invocation of `TracingHelper.logError(Span, Map)`

Reply:
Yes, I got that wrong about the error tag.
(My point was anyway more about the usage of the `logEntries` map, with the above mentioned alternatives being more straightforward and shorter. But, that may also be a matter of taste.)